File: api/src/main/java/cz/metacentrum/perun/spRegistration/common/models/Facility.java
package cz.metacentrum.perun.spRegistration.common.models;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.ObjectNode;
import cz.metacentrum.perun.spRegistration.Utils;
import cz.metacentrum.perun.spRegistration.common.enums.AttributeCategory;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Representation of Perun Facility.
 *
 * @author Dominik Frantisek Bucik <bucik@ics.muni.cz>
 */
@Getter
@Setter
@ToString
@EqualsAndHashCode(callSuper = true)
public class Facility extends PerunEntity {

	private String perunName;
	private String perunDescription;
	private Map<String, String> name = new HashMap<>();
	private Map<String, String> description = new HashMap<>();;
	private boolean oidc;
	private boolean saml;
	private Long activeRequestId;
	private boolean testEnv;
	private boolean editable = false;
	private List<User> admins = new ArrayList<>();
	private Map<AttributeCategory, Map<String, PerunAttribute>> attributes = new HashMap<>();

	public Facility(Long id) {
		super(id);
	}

	public Facility(Long id, String perunName, String perunDescription) {
		super(id);
		this.perunName = perunName;
		this.perunDescription = perunDescription;
	}

	/**
	 * Convert object to JSON representation
	 * @return JSON String
	 */
	public JsonNode toJson() {
		ObjectNode res = JsonNodeFactory.instance.objectNode();
		if (this.getId() == null) {
			res.set("id", JsonNodeFactory.instance.nullNode());
		} else {
			res.put("id", getId());
		}
		res.put("name", perunName);
		res.put("description", perunDescription);
		res.put("beanName", "facility");

		return res;
	}

	/**
	 * Convert JSON from perun to facility object
	 * @param json JSON from Perun
	 * @return Facility object or null
	 */
	public static Facility fromPerunJson(JsonNode json) {
		if (Utils.checkParamsInvalid(json)) {
			throw new IllegalArgumentException(Utils.GENERIC_ERROR_MSG);
		}

		Long id = json.get("id").asLong();
		String name = json.get("name").textValue();
		String description = json.hasNonNull("description") ? json.get("description").textValue() : null;

		return new Facility(id, name, description);
	}

}

File: gui/src/app/core/models/Facility.ts
import { User } from './User';
import { AttributesEntity } from './AttributesEntity';

export class Facility extends AttributesEntity {
  constructor(item: any) {
    super(item.attributes);
    if (!item) {
      return;
    }
    this.id = item.id;
    this.perunName = item.perunName;
    this.perunDescription = item.perunDescription;
    this.name = new Map<string, string>();
    if ('name' in item && item.name) {
      for (const k of Object.keys(item.name)) {
        this.name.set(k.toLowerCase(), item.name[k]);
      }
    }

    this.description = new Map<string, string>();
    if ('description' in item && item.description) {
      for (const k of Object.keys(item.description)) {
        this.description.set(k.toLowerCase(), item.description[k]);
      }
    }

    this.activeRequestId = item.activeRequestId;
    this.environment = item.environment;
    this.protocolUsed = item.protocol;

    this.managers = [];
    if (item.managers) {
      item.managers.forEach(user => {
        this.managers.push(new User(user));
      });
    }
    this.deleted = item.deleted;
  }

  id: number;
  perunName: string;
  perunDescription: string;
  name: Map<string, string>;
  description: Map<string, string>;
  environment: string;
  protocolUsed: string;
  activeRequestId: number;
  managers: User[];
  deleted: boolean;
}
